### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Shift): functors from a category to a category with a shift

### PR DESCRIPTION
**Pull request by joelriou**
Given a category `C`, and a category `D` equipped with a shift by a monoid `A`, we define a structure `SingleFunctors C D A` which contains the data of functors `functor a : C ⥤ D` for all `a : A` and isomorphisms `shiftIso n a a' h : functor a' ⋙ shiftFunctor D n ≅ functor a` whenever `n + a = a'`. These isomorphisms should satisfy certain compatibilities with respect to the shift on `D`.

---

When the derived category enters mathlib, this shall be used in order to redefine `Ext`-groups.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Shift/SingleFunctors.lean [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  /-- a family of functors `C ⥤ D` indexed by the elements of the additive monoid `A` -/
+  functor (a : A) : C ⥤ D
+  /-- the isomorphism `functor a' ⋙ shiftFunctor D n ≅ functor a` when `n + a = a'` -/
+  shiftIso (n a a' : A) (ha' : n + a = a') : functor a' ⋙ shiftFunctor D n ≅ functor a
+  shiftIso_zero (a : A) :
+    shiftIso 0 a a (zero_add a) = isoWhiskerLeft _ (shiftFunctorZero D A)
+  shiftIso_add (n m a a' a'' : A) (ha' : n + a = a') (ha'' : m + a' = a'') :
+    shiftIso (m + n) a a'' (by rw [add_assoc, ha', ha'']) =
+      isoWhiskerLeft _ (shiftFunctorAdd D m n) ≪≫ (Functor.associator _ _ _).symm ≪≫
+        isoWhiskerRight (shiftIso m a' a'' ha'') _ ≪≫ shiftIso n a a' ha'
```
Comment (chrisflav) [inline Mathlib/CategoryTheory/Shift/SingleFunctors.lean:47]:
```suggestion
  shiftIso (n a a' : A) (ha' : n + a = a') : functor a' ⋙ shiftFunctor D n ≅ functor a
  /-- `shiftIso 0` is the identity up to `shiftFunctorZero`. -/
  shiftIso_zero (a : A) :
    shiftIso 0 a a (zero_add a) = isoWhiskerLeft _ (shiftFunctorZero D A)
  /-- `shiftIso (m + n)` is the composition of `shiftIso n` with `shiftIso m` up
  to `shiftFunctorAdd`. -/
  shiftIso_add (n m a a' a'' : A) (ha' : n + a = a') (ha'' : m + a' = a'') :
    shiftIso (m + n) a a'' (by rw [add_assoc, ha', ha'']) =
      isoWhiskerLeft _ (shiftFunctorAdd D m n) ≪≫ (Functor.associator _ _ _).symm ≪≫
        isoWhiskerRight (shiftIso m a' a'' ha'') _ ≪≫ shiftIso n a a' ha'
```
Maybe some description of the compatibility conditions?

File: Mathlib/CategoryTheory/Shift/SingleFunctors.lean [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+@[simp, reassoc]
+lemma comp_hom (f : F ⟶ G) (g : G ⟶ H) (a : A) : (f ≫ g).hom a = f.hom a ≫ g.hom a := rfl
+
+@[ext]
+lemma hom_ext (f g : F ⟶ G) (h : f.hom = g.hom) : f = g := Hom.ext f g h
+
+/-- Constructor for isomorphisms in `SingleFunctors C D A`. -/
+@[simps]
+def isoMk (iso : ∀ a, (F.functor a ≅ G.functor a))
```
Comment (chrisflav) [inline Mathlib/CategoryTheory/Shift/SingleFunctors.lean:148]:
```suggestion

/-- Construct an isomorphism in `SingleFunctors C D A` by giving
level-wise isomorphisms and checking compatibility only in the forward direction. -/
@[simps]
def isoMk (iso : ∀ a, (F.functor a ≅ G.functor a))
```

File: Mathlib/CategoryTheory/Shift/SingleFunctors.lean [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    (comm : ∀ (n a a' : A) (ha' : n + a = a'), (F.shiftIso n a a' ha').hom ≫ (iso a).hom =
+      whiskerRight (iso a').hom (shiftFunctor D n) ≫ (G.shiftIso n a a' ha').hom) :
+    F ≅ G where
+  hom :=
+    { hom := fun a => (iso a).hom
+      comm := comm }
+  inv :=
+    { hom := fun a => (iso a).inv
+      comm := fun n a a' ha' => by
+        dsimp only
```
Comment (chrisflav) [inline Mathlib/CategoryTheory/Shift/SingleFunctors.lean:158]:
```suggestion
```

File: Mathlib/CategoryTheory/Shift/SingleFunctors.lean [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+instance (f : F ⟶ G) [IsIso f] (n : A) : IsIso (f.hom n) :=
+  (inferInstance : IsIso ((evaluation C D n).map f))
+
+variable (F)
+
+/-- The postcomposition `SingleFunctors C D A → SingleFunctors C E A` induced
+by a functor `D ⥤ E` which commutes with the shift by `A`. -/
+@[simps! functor shiftIso_hom_app shiftIso_inv_app]
+def postcomp (G : D ⥤ E) [G.CommShift A] :
+    SingleFunctors C E A where
```
Comment (chrisflav) [inline Mathlib/CategoryTheory/Shift/SingleFunctors.lean:199]: How is this used most often? Do you ever have a fixed `G` and consider the induced map  `SingleFunctors C D A → SingleFunctors C E A` (the doc string seems to suggest this)? If yes, the order of the arguments should maybe be switched such that `G` comes first?

Comment (joelriou) [reply Mathlib/CategoryTheory/Shift/SingleFunctors.lean:202]: Thanks for the suggestions. I have changed the docstring.

File: Mathlib/CategoryTheory/Shift/SingleFunctors.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Shift/SingleFunctors.lean]: I have added docstrings.

**Review by chrisflav — APPROVED** [review]
Looks good to me.

**Comment by erdOne** [comment]
There should be a linkifier on Zulip to link to approved PRs or I'll forget to look at them.
Thanks! (And also thanks Chris for reviewing this)
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**